src/herder/Upgrades.cpp

```diff
@@ -1029,4 +1242,201 @@ Upgrades::applyReserveUpgrade(AbstractLedgerTxn& ltx, uint32_t newReserve)
         prepareLiabilities(ltx, header);
     }
 }
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+ConfigUpgradeSetFrameConstPtr
+ConfigUpgradeSetFrame::makeFromKey(AbstractLedgerTxn& ltx,
+                                   ConfigUpgradeSetKey const& key)
+{
+    auto ltxe = ltx.loadWithoutRecord(ConfigUpgradeSetFrame::getLedgerKey(key));
+    if (!ltxe)
+    {
+        return nullptr;
+    }
+    auto const& contractData = ltxe.current().data.contractData();
+    if (contractData.val.type() != SCV_BYTES)
+    {
+        return nullptr;
+    }
+    auto const& bytes = contractData.val.bytes();
+
+    ConfigUpgradeSet upgradeSet;
+    try
+    {
+        xdr::xdr_from_opaque(bytes, upgradeSet);
+    }
+    catch (xdr::xdr_runtime_error&)
+    {
+        return nullptr;
+    }
+
+    return std::shared_ptr<ConfigUpgradeSetFrame>(
+        new ConfigUpgradeSetFrame(upgradeSet, key));
+}
+
+ConfigUpgradeSetFrame::ConfigUpgradeSetFrame(
+    ConfigUpgradeSet const& upgradeSetXDR, ConfigUpgradeSetKey const& key)
+    : mConfigUpgradeSet(upgradeSetXDR)
+    , mKey(key)
+    , mValidXDR(isValidXDR(upgradeSetXDR, key))
+{
+}
+
+bool
+ConfigUpgradeSetFrame::isValidXDR(ConfigUpgradeSet const& upgradeSetXDR,
+                                  ConfigUpgradeSetKey const& key) const
+{
+    if (key.contentHash != sha256(xdr::xdr_to_opaque(upgradeSetXDR)))
+    {
+        CLOG_DEBUG(Herder,
+                   "Got bad configUpgradeSet. Does not match hash in key {}",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+
+    if (upgradeSetXDR.updatedEntry.empty())
+    {
+        CLOG_DEBUG(Herder, "Got bad configUpgradeSet {}: no entries updated",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+
+    if (!std::is_sorted(
+            upgradeSetXDR.updatedEntry.begin(),
+            upgradeSetXDR.updatedEntry.end(),
+            [](ConfigSettingEntry const& a, ConfigSettingEntry const& b) {
+                return a.configSettingID() < b.configSettingID();
+            }))
+    {
+        CLOG_DEBUG(Herder,
+                   "Got bad configUpgradeSet {}: the entries are not ordered",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+    if (std::adjacent_find(
```

Comment:
`is_sorted` above already does not allow for duplicates

Reply:
That's not true - it 'Checks if the elements in range [first, last) are sorted in non-descending order'

Reply:
bleh yes! I always get is_sorted wrong

So I guess the suggestion becomes: as we use a custom predicate we should just use `>=` for `adjacent_find`  which should do what we want (can't use this with `is_sorted` as `>=` is not `Compare`).